### ISSUE TITLE
Exorcise assertions from Scylla

### ISSUE THREAD

**Issue by nyh** [scylladb/scylladb#7871]
In less than a month, I found two examples (issue #7740 and issue #7825) of valid CQL requests which trigger an assertion failure in Scylla and accordingly, a crash. Issue #6995 is a similar (though more elaborate) crash I found a few months ago. 

Those are bugs like all normal bugs - the CQL requests in question do not do what they should be doing. 
However, the assertion failures make them worse than normal bugs:

1. When a user tries a new type of query and it fails, the user will probably view this as much less dramatic event than the case that the cluster or some of its nodes come crashing down. In most cases, the user would not even know **why** the nodes came down - it may not be obvious that it happened because of some specific request, until Scylla experts will analyze the logs or core dumps to find the reason for the assertion failure.
2. Asserts are dangerous if in the future we want to support multi-tenancy. They allow one tenant to crash the entire system just by submitting maliciously crafted CQL requests.

We have (in Seastar) the function `on_internal_error` which aims to solve this problem - by throwing an exception instead of crashing, except when the user deliberately configures Scylla to crashes on bugs (i.e., in test runs). We should audit the code for any remaining asserts and replace them by `on_internal_error`.

We should consider whether there are places in Scylla where it is better to still have a crash instead of an exception. Perhaps when noticing a disaster which makes any more progress on anything impossible. Or perhaps there are cases that we believe that nothing besides a coredump would be able to discover the cause for some unexpected event (I think @avikivity once said that). But in those cases, if they exist, we should add comments why we decided to use an assertion instead of `on_internal_error`.

P.S. The use of the verb **to exorcise** in the title of the issue, instead of **to excise**, was deliberate, not a typo :-)

**Comment by nyh** [comment]
> In less than a month, I found two examples (issue #7740 and issue #7825) of valid CQL requests which trigger an assertion failure in Scylla and accordingly, a crash. Issue #6995 is a similar (though more elaborate) crash I found a few months ago.
> 
> Those are bugs like all normal bugs - the CQL requests in question do not do what they should be doing.
> However, the assertion failures make them worse than normal bugs:
> 
>     1. When a user tries a new type of query and it fails, the user will probably view this as much less dramatic event than the case that the cluster or some of its nodes come crashing down. In most cases, the user would not even know **why** the nodes came down - it may not be obvious that it happened because of some specific request, until Scylla experts will analyze the logs or core dumps to find the reason for the assertion failure.
> 
>     2. Asserts are dangerous if in the future we want to support multi-tenancy. They allow one tenant to crash the entire system just by submitting maliciously crafted CQL requests.
> 
> 
> We have (in Seastar) the function `on_internal_error` which aims to solve this problem - by throwing an exception instead of crashing, except when the user deliberately configures Scylla to crashes on bugs (i.e., in test runs). We should audit the code for any remaining asserts and replace them by `on_internal_error`.
> 
> We should consider whether there are places in Scylla where it is better to still have a crash instead of an exception. Perhaps when noticing a disaster which makes any more progress on anything impossible. Or perhaps there are cases that we believe that nothing besides a coredump would be able to discover the cause for some unexpected event (I think @avikivity once said that). But in those cases, if they exist, we should add comments why we decided to use an assertion instead of `on_internal_error`.

P.S. The use of the verb _to exorcise_ in the title of the issue, instead of _to excise_, was deliberate, not a typo :-)

**Comment by avikivity** [comment]
This issue can never be marked as fixed, so it is doomed to be forever in the backlog. Any issue which references the entire code base is going to become part of the background radiation of issues.

Try to file fixable issues.

**Comment by nyh** [comment]
I had exactly the same argument with @dekimir today about something else (bugs in reversed types).
My theory is that when you encounter a bug which you **know** is systemic - it is a trap which the code fell into more than once - you need to address that systemic bug. Not just to pretend it's enough to fix the one case that bothered you most today and leave all the rest.

For **this** issue, this bug can be marked closed once either all calls to assert() are gone, or all are reviewed and considered why they need to remain asserts (preferably leaving behind a comment explaining why). That's a lot of work, but does not need to be "forever".

Yes, I could have not filed this bug, and we can have one less bug. That would not make it go away.
Another benefit of having a bug is that it's a "proof" that it needs to be fixed. Next month, when someone submits a patch adding a new assert() to the code we can tell him: you can't do this, this is a bug - check out issue 7871.

**Comment by nyh** [comment]
Today we had another example of the importance of this issue - a user encountered a bug in Scylla's handling of a very specific query, but the code in question (in an old branch) had an asssert(), which led to the entire cluster crashing together. The user didn't even know which query caused this crash, because none of the queries ever reported an error.

The specific assert() that caused this is gone from modern code, but the general problem of assert remains. Let's use throw or on_internal_error() everywhere, not assert().

**Comment by elcallio** [comment]
I wish we had asserts as god meant them: extensive (potentially expensive) checks, only for debug/check mode. The fact that we have asserts compiled into release runtime makes it prohibitive to use them when the checks would be even somewhat costly. I would like to be able to sprinkle various sanity check asserts into code, knowing they will only affect debug test runs or similar. 
Actual, runtime, checks for broken/uncontinuable conditions should as you say be various forms of ensure/internal_error.

**Comment by nyh** [comment]
By the way as I discovered yesterday in #12815, that we have another bad habit as bad as `assert()` - and it is calling `abort()` in various places. #12815 shows three similar queries which can be used to immediately crash Scylla because our code decided to abort Scylla instead of just the failed operation. 

We need to exorcise these abort() calls as well. Needless to say, allowing a bad query to crash Scylla was always a bad idea, but will become an even worse idea when we want to support multi-tenancy.

**Comment by mykaul** [comment]
This is a good item for the quality improvement effort. @nyh - any idea how to split it to multiple items, bugs with owners? Perhaps grep for all places it happens, open one item per occurrence. . .?

**Comment by denesb** [comment]
> This is a good item for the quality improvement effort. @nyh - any idea how to split it to multiple items, bugs with owners? Perhaps grep for all places it happens, open one item per occurrence. . .?

I think this is pretty easy to split. Everybody can pick a file (or more) or submodule and focus on those. Go over each assert and remove them, replace with throw or `on_internal_error()` as appropriate.

There is plenty for everyone here:
```
$ ag assert --ignore=seastar --ignore=test --ignore=tools --ignore=swagger-ui | wc -l
1033
```

**Comment by mykaul** [comment]
With slightly more filtering, we still get a lot of items...
```
[ykaul@ykaul scylladb]$ ag "assert\(" --ignore=seastar --ignore=test --ignore=tools --ignore=swagger-ui --ignore=abseil --ignore=libdeflate|grep -v RAPIDJSON |wc -l
927
```

Need to see how we get to something more reasonable and actionable...

**Comment by denesb** [comment]
> Need to see how we get to something more reasonable and actionable...

I think the only way this will work is, as a long-term project. No one would be able to do this in anything close to a single PR, especially because after a few hours of scanning through asserts, one will need a break from it and do something else.

So the way I see this working is the same way our effort to remove our globals worked. A low-priority side-project running for months with low-intensity, grappling the daunting task one small PR at a time. This can be done by one or more individuals.

**Comment by mykaul** [comment]
Per an internal discussion, I'll split the efforts around CQL (frontend) to its own issue - there's more concern on assertions and their impact there.

**Comment by nyh** [comment]
> This is a good item for the quality improvement effort. @nyh - any idea how to split it to multiple items, bugs with owners? Perhaps grep for all places it happens, open one item per occurrence. . .?

No need to split this issue - I think quite the contrary, we need one person to look over the entire code base, look at every instance of assert() and abort(), and determine a strategy on how to avoid them - or when *not* to avoid them. I counted them now and there are about 500 of them. Not something that a single person can't do in a week (especially since a lot of times you have many similar asserts in the same file). Alternatively, we can split the work according to subdirectories, and so on, if several people want to work on this.

I'm not working on this myself. I wanted to triage the issues I opened (more than 200, including this one),  but my manager asked me to do something else so that's what I'm doing.

**Comment by nyh** [comment]
> With slightly more filtering, we still get a lot of items...
> 
> ```
> [ykaul@ykaul scylladb]$ ag "assert\(" --ignore=seastar --ignore=test --ignore=tools --ignore=swagger-ui --ignore=abseil --ignore=libdeflate|grep -v RAPIDJSON |wc -l
> 927
> ```
> 
> Need to see how we get to something more reasonable and actionable...

Don't forget to exclude static_assert which is good. I counted 495 in .cc files and 297 in .hh files.
Counting abort() is more difficult because we have an unrelated abort() method in various classes.

**Comment by mykaul** [comment]
I've opened https://github.com/scylladb/scylladb/issues/13970 on CQL - I don't see how a single person can look at all of the issues, but let's see how it goes with CQL.

**Comment by nyh** [comment]
Turns out some of these assert() calls can even be hiding in Seastar :-( @nuivall discovered a crash due to an assertion in Seastar in https://github.com/scylladb/scylladb/issues/14403#issuecomment-1611722033. 
This assertion probably indicates a real bug, which is important to know about, but it would have still have been nicer to get a more regular error report, and not the entire Scylla node (or even cluster) coming down.

**Comment by mykaul** [comment]
> Turns out some of these assert() calls can even be hiding in Seastar :-( @nuivall discovered a crash due to an assertion in Seastar in [#14403 (comment)](https://github.com/scylladb/scylladb/issues/14403#issuecomment-1611722033). This assertion probably indicates a real bug, which is important to know about, but it would have still have been nicer to get a more regular error report, and not the entire Scylla node (or even cluster) coming down.

Perhaps we should open a similar meta-issue in Seastar, with the hope we'll get to somehow work on those as well.

**Comment by kbr-scylla** [comment]
@nyh could you please create a PR with a document (in doc/dev/) with proposed guidelines to using `assert` vs `on_internal_error` vs `on_fatal_internal_error`?

I've already had multiple situations where a patch tried to introduce an `assert`, I requested to it to `on_internal_error`, but got push-back because there are no clear guidelines and people don't want to change their habits "just because of someone's personal vendetta against `assert`" (paraphrasing).

We need a coding guideline where we can point people to prevent new `assert`s from being introduced. Otherwise this issue will never be solved even if we went through the entire codebase now and replaced every instance of `assert`.

**Comment by nyh** [comment]
> Turns out some of these assert() calls can even be hiding in Seastar :-( @nuivall discovered a crash due to an assertion in Seastar in [#14403 (comment)](https://github.com/scylladb/scylladb/issues/14403#issuecomment-1611722033). 

Another important observation that was made in #14403 is that when we have an `assert()` failure, in either Scylla or Seastar, its text isn't printed using Seastar's logging mechanism but rather gets printed directly to stdout (printed by the standard C function `assert()`), so users who set logging to syslog do not see the text of the assertion failure! This made the user in #14403 not even know why Scylla crashed, until only the core dump generated by the `assert()` could confirm which assert it was that failed.

So that's yet another reason we should never ever call the C function assert() in Scylla, and use something else. As if we needed more reasons.

**Comment by elcallio** [comment]
No, that is a reason why not o use assert as a production error mechanism. It is not. Assert:s should be for errors that never should be allowed to reach production code. They are for catching programmer logical fallacies, such as changing the conditions under which something else is invoked, etc. They are not for runtime validation etc, they are (debug-only), extra checks to verify assumptions on how the world is structured. Anything that concerns things that actually _could_ (by the definitions of the problem space) happen should be a proper runtime check and "real" errors, indeed. 
But assert literally means "break into debugger now, and please call me there asap because the world is broken".

Also, one can (and we do) redefine exactly what and when happens in an assert. Prints can be redirected etc. They should not. But they can.

Seriously, we should have _more_ asserts. But stop using asserts as a quick-and-dirty runtime check + no actual error handling. 
I want to have _loads_ of asserts that never execute in release mode (so they can be potentially intrusive), just to ensure that pre- and postconditions comply with my view of the world at the time of writing. And consequently these failing can tell us when someone has broken said assumptions _early_ before they reach production code.

**Comment by kbr-scylla** [comment]
`on_internal_error` is exactly for the purposes you listed @elcallio - that is, catching conditions which should **never** happen (if they happen, it's a bug) - programmer logical fallacies. It's supposed to be a (better) alternative for `assert`.

For runtime checks for user errors - conditions which _may_ happen - we have `throw std::runtime_error` and so on.

**Comment by elcallio** [comment]
No, on_internal_error is for an error that _can_ happen, but for which there is no handling. There is a difference. Mainly that the check is maintained in production code.

Assert is for non-production code, it is literally to put up a nice big dialog saying "this condition failed, would you like to debug". When doing debug-runs on a dev machine.

I agree that scylla uses asserts wrong. But I want to use them more, and right.

**Comment by kbr-scylla** [comment]
> No, on_internal_error is for an error that can happen, but for which there is no handling.

Source for that claim?
Let's see what seastar git log has to say about it.
```
core: Add a on_internal_error helper

In scylla there are places where we want to check an invariant, but
not bring down the database if it fails.

Like the invariants checked with assert, these are things that should
never happen and indicate bugs, not just unusual issues like a
corrupted file.

Unlike the invariants checked with assert, we are reasonably confident
that it is possible to abort a particular task without bringing down
the entire server.

To handle that, scylla has a on_internal_error function that will
either throw or abort depending on a runtime flag.

This patch moves on_internal_error to seastar. It seems general enough
and we want this flag to be always set to abort during tests, which is
more naturally done in seastar.
```

**Like the invariants checked with assert, these are things that should
never happen and indicate bugs, not just unusual issues like a
corrupted file.**

> There is a difference. Mainly that the check is maintained in production code.

Whether the check is maintained in production code or not is orthogonal to whether this is an error which cannot or can happen. There is no implication.

My opinion is that for conditions which should never happen in production we should also have checks in release mode. And, given my understanding of the purpose of `on_internal_error`, we should use it for these checks.

**Comment by elcallio** [comment]
Given that code that code calling on_internal_error treats it like a noreturn throw, and that there is no one "catching" said internal error, I would say that it very much conforms to "no handling". 

The problem with checks in release mode is that it limits the amount of checking feasible. A check that is computationally expensive will not be done, because the overhead does not match the expected return. An assert can be super expensive, because we are in "correctness checking" mode, not production/benchmark. 

Again, all the arguments against the _current_ use of asserts in much of the code is fully correct. But on the other hand there are not _enough_ asserts either, simply because people are reluctant to add checks of things that are "obvious" etc. Yet, having those checks for debug mode might help finding, pinpointing and make debugging of easier, all those pesky cases of mats being torn out from under feet that keep happening across the ecosystem.

**Comment by kbr-scylla** [comment]
> Given that code that code calling on_internal_error treats it like a noreturn throw, and that there is no one "catching" said internal error, I would say that it very much conforms to "no handling".

Same as `assert`, there is also "no handling" for `assert`, I'm not disputing that.

I'm disputing this part of your claim: "No, on_internal_error is for an error that can happen"
The evidence I gave proves that `on_internal_error` is for an error that **cannot** happen, exactly like `assert`.

**Comment by elcallio** [comment]
If the check is in production code, by definition it can happen. :-)
Lets restate it then to be for "things that cannot happen that should be guarded against even after we've exited (internal) testing".

**Comment by nyh** [comment]
> They are not for runtime validation etc, they are (debug-only)

In some traditions, release builds are compiled with the NDEBUG macro which removes any assert calls from the code. Scylla does not, and never did, adhere to this tradition - all our builds **do not** define NDEBUG, so asserts are included in all builds, including release builds. We can change this, but we never did.

But note that in many cases it (using NDEBUG as intended) wouldn't have helped, or even made things worse. If we have assert(...) somewhere, it usually means the code below won't work if the assertion isn't true. The code below might even crash if the assertion isn't true. So just having NDEBUG make the assert() be ignored doesn't help prevent crashes in release builds! Instead, `on_internal_error()` does exactly the right things:

1. In debug builds (and optionally, also in release builds - we use this option in Jenkins), it causes a crash, just like `assert()`. The developer can't possibly miss that.
2. In release builds, it causes an exception. This stops the faulty code (the code directly following the assert()) from continuing to run, but possibly doesn't crash the entire Scylla. For example, maybe we have a bug, even a serious bug, in handling a SELECT, and the exception will ruin that request, but other requests will continue to work just fine - and we prevented a small bug from turning into a disaster.

Calle, if you still think that assert()s in Scylla aren't a disaster, I ask you to please take a look at https://github.com/scylladb/scylladb/issues/14403. In that issue we had a **real bug** - Scylla forgot to cleanly close the socket when a user's HTTP connection broke. But instead of this becoming a serious log message and minor nuisance (at worse, perhaps some sort of slow resource leak), this became an immediate crash of the entire cluster (!), repeatedly, and lost data. I don't see any way to explain why this is a good thing.

> Also, one can (and we do) redefine exactly what and when happens in an assert. Prints can be redirected etc. They should not. But they can.

What do you mean "and we do" - do we redefine assert()? It's a standard C function...
If we redefine assert to be on_internal_error, we can use assert ;-) 

> I want to have _loads_ of asserts that never execute in release mode (so they can be potentially intrusive), just to ensure that pre- and postconditions comply with my view of the world at the time of writing.

Please see my comment about on NDEBUG. This isn't the case in Scylla today - those asserts **do** get executed in release mode. And as I noted above, it's **not a great idea** to simply not execute them in release mode, because if those assertions failed, the code below them can often crash, so we replaced one crash (assert()) by another, and that's not our goal.

**Comment by kbr-scylla** [comment]
> If the check is in production code, by definition it can happen. :-)

Sorry, I don't get this logic. Bugs do happen in production unfortunately, and protections against them are IMO necessary.
Similarly Linux has the `BUG` and `BUG_ON` macros which are enabled in release IIRC.

>  Lets restate it then to be for "things that cannot happen that should be guarded against even after we've exited (internal) testing".

Agreed.
And most of our uses of `assert` -- IMO -- should be turned into such checks that should be enabled even in production.

The super-expensive checks you're talking about -- sure, let's have those only in debug mode / during testing. But we don't really have those. (Well, we do have error injections which are compiled away in release, but that's something a bit different.)

Most of the checks we're checking with `assert` are super cheap (like checking for `nullopt` or `nullptr`). We can easily turn them into `on_internal_error`.

**Comment by nyh** [comment]
> The problem with checks in release mode is that it limits the amount of checking feasible. A check that is computationally expensive will not be done, because the overhead does not match the expected return. An assert can be super expensive, because we are in "correctness checking" mode, not production/benchmark.

We don't need assert() for that. For this we have `#ifdef SEASTAR_DEBUG`, with which you can wrap not only extra checks, but also extra class members, and so on, needed to support these extra checks. Having `assert()` disappear in release mode (which, as I noted above, **does not happen** in Scylla's build system!) is a rather weak substitute for that. 


> Again, all the arguments against the _current_ use of asserts in much of the code is fully correct. But on the other hand there are not _enough_ asserts either, simply because people are reluctant to add checks of things that are "obvious" etc. Yet, having those checks for debug mode might help finding, pinpointing and make debugging of easier, all those pesky cases of mats being torn out from under feet that keep happening across the ecosystem.

I agree. But that's not an excuse to use specifically the `assert()` function, any more than a complaint that we don't have enough logging means that we must use the `printf()` function. I agree that the `on_internal_error()` function is not a direct substitute for `assert()` (it requires an if and on_internal_error, three lines of code). Maybe we should create a macro `on_internal_error_assert(logger, expression)` that has the same API as assert - it checks the given expression and uses the string for of the expression (`#expression`) for the assertion message.

**Comment by elcallio** [comment]
> > They are not for runtime validation etc, they are (debug-only)
> 
> In some traditions, release builds are compiled with the NDEBUG macro which removes any assert calls from the code. Scylla does not, and never did, adhere to this tradition - all our builds **do not** define NDEBUG, so asserts are included in all builds, including release builds. We can change this, but we never did.
> 
> But note that in many cases it (using NDEBUG as intended) wouldn't have helped, or even made things worse. If we have assert(...) somewhere, it usually means the code below won't work if the assertion isn't true. The code below might even crash if the assertion isn't true. So just having NDEBUG make the assert() be ignored doesn't help prevent crashes in release builds! Instead, `on_internal_error()` does exactly the right things:
> 

Again, I am not arguing against the issues you've listed, nor that scylla is f**bared w.r. to assert usage. I am only arguing that I want proper, release -> -DNDEBUG asserts to use as _extra_ guards, for things that do not qualify as "proper" check -> internal_error().

>     1. In debug builds (and optionally, also in release builds - we use this option in Jenkins), it causes a crash, just like `assert()`. The developer can't possibly miss that.
> 
>     2. In release builds, it causes an exception. This stops the faulty code (the code directly following the assert()) from continuing to run, but possibly doesn't crash the entire Scylla. For example, maybe we have a bug, even a serious bug, in handling a SELECT, and the exception will ruin that request, but other requests will continue to work just fine - and we prevented a small bug from turning into a disaster.

That is the opposite of what you want with a proper assert. An assert is for things that should always stop the process, and for which you grab your debugger and check what is wrong. An assert-level check that translates into a throw (vulnerable to catch(...)) will only further contaminate the system anyway. But that is beside the point. Again, an assert is for things that help the developer ensure his code is doing what he expected under the expected conditions. It is a dev tool. Not a production error check tool.

For anything relating to checks that should be in runtime, I agree with you. But I want asserts that work like asserts always have worked in any sane code base. I.e. trigger only in debug mode, invisible otherwise. 

One is not a replacement for the other. They are different things. 

> 
> 
> Calle, if you still think that assert()s in Scylla aren't a disaster, I ask you to please take a look at #14403. In that issue we had a **real bug** - Scylla forgot to cleanly close the socket when a user's HTTP connection broke. But instead of this becoming a serious log message and minor nuisance (at worse, perhaps some sort of slow resource leak), this became an immediate crash of the entire cluster (!), repeatedly, and lost data. I don't see any way to explain why this is a good thing.
> 

It is not. It is asserts used wrong, obviously. I have hated the way asserts are in scylla since we started, since I can't assert as much as I would like to. But I also don't consider them an alternative to check + throw. 

> > Also, one can (and we do) redefine exactly what and when happens in an assert. Prints can be redirected etc. They should not. But they can.
> 
> What do you mean "and we do" - do we redefine assert()? It's a standard C function... If we redefine assert to be on_internal_error, we can use assert ;-)

Any sane c runtime lets you hook assertion handlers. If you really want, yes, you can turn them into internal error. But that would be incredibly bad. Because assert and "abort this task, maybe, if possible" are not the same thing. 

> 
> > I want to have _loads_ of asserts that never execute in release mode (so they can be potentially intrusive), just to ensure that pre- and postconditions comply with my view of the world at the time of writing.
> 
> Please see my comment about on NDEBUG. This isn't the case in Scylla today - those asserts **do** get executed in release mode. And as I noted above, it's **not a great idea** to simply not execute them in release mode, because if those assertions failed, the code below them can often crash, so we replaced one crash (assert()) by another, and that's not our goal.

If the issue needs to be checked in release mode, then it is not an assert. 
There is a reason state_assert was added to c++14. Developers want condition checking during code development, and assert is just the level above compile time. It is initial testing time. It is testing checks for when breaking into a debugger is a viable option. 

Scylla seems to have decided that debugging in general is pointless, since what we call a "dev" build is literally without debug symbols, and even our debug mode runs with opts that kill more than half of debug info (vars). But that is how you get ants (bugs).

**Comment by elcallio** [comment]
> > If the check is in production code, by definition it can happen. :-)
> 
> Sorry, I don't get this logic. Bugs do happen in production unfortunately, and protections against them are IMO necessary. Similarly Linux has the `BUG` and `BUG_ON` macros which are enabled in release IIRC.

Ok, then try to think on it like this: if (cond) internal_error() is a check for when code is run. Assert is a check for _me_ when the code is being developed. It assumes this is a check not needed for the field. It is a check for testing conditions only. 

In the old days, a code base would typically have ASSERT(...) and RUNTIME_ASSERT(...). The latter being asserts that should always be present. Aka on_internal_error stuff.

**Comment by elcallio** [comment]
> > Again, all the arguments against the _current_ use of asserts in much of the code is fully correct. But on the other hand there are not _enough_ asserts either, simply because people are reluctant to add checks of things that are "obvious" etc. Yet, having those checks for debug mode might help finding, pinpointing and make debugging of easier, all those pesky cases of mats being torn out from under feet that keep happening across the ecosystem.
> 
> I agree. But that's not an excuse to use specifically the `assert()` function, any more than a complaint that we don't have enough logging means that we must use the `printf()` function. I agree that the `on_internal_error()` function is not a direct substitute for `assert()` (it requires an if and on_internal_error, three lines of code). Maybe we should create a macro `on_internal_error_assert(logger, expression)` that has the same API as assert - it checks the given expression and uses the string for of the expression (`#expression`) for the assertion message.

Or you could just change the uses of assert that are runtime checks to be runtime checks, and then use assert as god intended?

**Comment by kbr-scylla** [comment]
> Ok, then try to think on it like this: if (cond) internal_error() is a check for when code is run. Assert is a check for me when the code is being developed. It assumes this is a check not needed for the field. It is a check for testing conditions only.

> In the old days, a code base would typically have ASSERT(...) and RUNTIME_ASSERT(...). The latter being asserts that should always be present. Aka on_internal_error stuff.


@elcallio looks like we're running two parallel discussions.

The role of `assert` in Scylla **currently** **is** what you called `RUNTIME_ASSERT` -- because as Nadav said, we don't disable asserts in production.

What you called `ASSERT`, i.e. testing-only asserts, we have in the form of `#ifdef SEASTAR_DEBUG`.

Now given the current state of Scylla, the case of replacing `assert`s with `on_internal_error`, I think is clear.

A parallel discussion is whether we should introduce more testing-only checks, i.e. those "real" `ASSERT`s that you'd like to see more of.

**Comment by kostja** [comment]
Assert in production as we use them today translate to std::logic_error. A violation of an invariant can be irrecoverable or recoverable. std::on_internal_error is for an irrecoverable failure. Our current asserts therefore sit on two chairs and this is why this argument is so heated.

I agree with Calle, we should leave asserts be for what they were made for - debug-only checks, for critical things use std::logic_error and std::on_internal_error.

**Comment by denesb** [comment]
I think the path of least resistance would be to introduce a new kind of `DEBUG_ASSERT()` which we can use thoroughly in the code, to check potentially expensive preconditions. Then go through the current `assert()`:s in scylla code and decide which should be converted to `DEBUG_ASSERT()` and which should be converted to `on_internal_error()` instead or even left as plain `assert()`.

**Comment by regevran** [comment]
With the adoption of contracts into C++26 we can now change our code in a way that this issue will eventfully be closed.
A discussion and agreements must be made - but a standard way to write and handle assertions is now (almost) available. 
As a first item to discuss:
The difference between _errors_ which are recoverable to _bugs_ which a human programmer should look at and fix.

**Comment by nyh** [comment]
> With the adoption of contracts into C++26 we can now change our code in a way that this issue will eventfully be closed.

If I understand correctly, these contracts are very similar to our `on_internal_error()`: Both mechanisms allow you configure globally whether a failed check will cause an immediate crash or an exception. Both mechanism allow to you, for example, ask for crashes during testing, but exceptions in production.

This issue is about getting rid of all **other** calls in our code to `assert()`, `SCYLLA_ASSERT()` or `abort()` - and keeping `on_internal_error()`, so in this respect it doesn't care if you use `on_internal_error()` or contracts. The problem with those other mechanisms (assert, abort, etc.) is that they causes crashes in production, which cause us nothing but grief (see all the examples I gave above). We could convert all these assertions to contracts, but honestly, we can do it today with `on_internal_error()` and don't need to wait for contracts to appear in 2026.

The contracts feature is perhaps a nicer and less verbose syntax for `on_internal_error()`, and we can start using it when it's ready, but I don't see how it gives us more power.

> A discussion and agreements must be made - but a standard way to write and handle assertions is now (almost) available. 
> As a first item to discuss: The difference between _errors_ which are recoverable to _bugs_ which a human programmer should look at and fix.

There's a slippery slope on what is considered "recoverable" and what do you actually mean when you claim that **bugs** are not recoverable.

As a random example, a person who is developing materialized views might think that the view-update code tries to write to a view and discovers the schema is corrupt (e.g., a column is missing in the view), then this is a "bug" and an "unrecoverable disaster" and think that we might as well crash Scylla. But the typical user of Scylla will not agree: They will prefer to see writes to this specific view failing instead of Scylla crashing, because:

1. The user will see the failed writes, and know which table has a problem and what it is.
2. The user will be able to continue to (say) read from the table, and to read or write to other tables which are not corrupt.

(see links above for real-world disasters that happened when users got assert()-caused crashes because Scylla detected a "bug" and sadly the programmer believed that to remove all doubts - Scylla should crash when it detects a bug).

In some case we really have unrecoverable errors when we know Scylla **must not continue** at all. An example is an IO error, after which we can't be sure about the state of **anything** done on this node, not just the operation which failed. But these should be very rare (in fact, as can be seen in https://github.com/scylladb/scylladb/issues/6819, even for IO errors it is not universally true). Almost all **bugs** are no unrecoverable in this sense - these bugs can abort just the operation that contains them (even if it was a very important operation) and allow other operations, or operations on other tables, and so on, to continue.

**Comment by nyh** [comment]
Last year, in commit aa1270a00c40af8dddb71b8b8d80d10957d92798, we changed all calls to assert() to SCYLLA_ASSERT(). This change is meaningless in regards to this issue - as SCYLLA_ASSERT() does exactly the same as our old assert() did and has exactly the same problem. 

So now I see people adding new SCYLLA_ASSERT() calls in code instead of assert(), causing exactly the same risks I'm trying to warn about. A quick "git grep" shows we now have 842 calls to SCYLLA_ASSERT() outside tests :-(

**Comment by regevran** [comment]
>  Almost all bugs are no unrecoverable in this sense - these bugs can abort just the operation that contains them

Contracts distinguishes between recoverable errors that are not bugs and bugs [2.1 What Are Contracts?, second paragraph](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2025/p2900r14.pdf).

Because _many_ systems behave like you described: can continue in other "threads" of execution while canceling the buggy one - it might be that contracts do not fit to such systems.

However, the ability to `throw` after a contract was violated is kind of a mix between the two idioms of an error and a bug: a bug is caught by a contract but execution continues with a thrown exception. It lets developers clearly indicate the expected correct conditions (contracts) while throwing if they were violated.

**Comment by nyh** [comment]
> So now I see people adding new SCYLLA_ASSERT() calls in code instead of assert(), causing exactly the same risks I'm trying to warn about. A quick "git grep" shows we now have 842 calls to SCYLLA_ASSERT() outside tests :-(

I still see people adding SCYLLA_ASSERT to new patches all the time. I'm starting to suspect that people prefer assert or SCYLLA_ASSERT over on_internal_error() just because of the convenience. I opened https://github.com/scylladb/scylladb/issues/28308 proposing to add an assert-like function that uses on_internal_error(), to make it the best of all worlds: Convenient to use as assert(), but doesn't crash user deployments like assert() does.

**Comment by bhalevy** [comment]
> > So now I see people adding new SCYLLA_ASSERT() calls in code instead of assert(), causing exactly the same risks I'm trying to warn about. A quick "git grep" shows we now have 842 calls to SCYLLA_ASSERT() outside tests :-(
> 
> I still see people adding SCYLLA_ASSERT to new patches all the time. I'm starting to suspect that people prefer assert or SCYLLA_ASSERT over on_internal_error() just because of the convenience. I opened [#28308](https://github.com/scylladb/scylladb/issues/28308) proposing to add an assert-like function that uses on_internal_error(), to make it the best of all worlds: Convenient to use as assert(), but doesn't crash user deployments like assert() does.

Can't we redefine SCYLLA_ASSERT to use on_internal_error_noexecpt instead of `__assert_fail`?

**Comment by avikivity** [comment]
It's misleading. ASSERT should be an assertion, not an on_internal_error. Each call site needs to be inspected and converted.

**Comment by bhalevy** [comment]
> It's misleading. ASSERT should be an assertion, not an on_internal_error. Each call site needs to be inspected and converted.

Edit: I meant `on_fatal_internal_error`, not `on_internal_error_noexcept`.
I.e. it has the same effect as a failed assert of aborting the program with a visible backtrace and coredump.

**Comment by bhalevy** [comment]
But I agree that every ASSERT should be vetted to see if it's indeed asserting an invariant that should never happened, and if it did, it is an internal error, or is it something that can be handled in production yet we do want to abort on it in testing time to understand how we got there.
Or, if it's a valid case that should be properly handled.

**Comment by m-szymon** [comment]
As one of the "people" I will explain my reasons and understanding.

> It's misleading. ASSERT should be an assertion, not an on_internal_error. 

Assertion that goes into production is already misleading. I am used to assert that:
1. Documents contract (code doesn't lie).
2. Helps testing/debbuging (Consequences are not always immediate, and could be missed in tests, even if the buggy path is covered).
3. Has no influence on production.

So assertions are good, not bad. Assertions used wrong (e.g. on user input - we want that checks in production) is bad, as everything.

Just in our case we want to be more defensive and pragmatic. Our code is big and complex, actively changed by many people. And test coverage is low. Assertions won't help us, because it won't be hit in testing. So we should prefer more production runtime checks.
So (assuming assert is disabled in production) right message wouldn't be "don't use assert", but "use assert, but consider using on_internal_error - it is almost always better."
Now if you really want assert in production you are facing question "do you really want to crash server?". And I bet that in vast majority the answer is "No or I don't know - I am writing my component, I don't see the whole system". And if you do, simple "SCYLLA_ASSERT" is not enough - it should explicitly say that it is killing production server with a comment explaining it from the system perspective. 

> I still see people adding SCYLLA_ASSERT to new patches all the time. I'm starting to suspect that people prefer assert or SCYLLA_ASSERT over on_internal_error() just because of the convenience.

Improving convenience is always good, but SCYLLA_ASSERT is used because 1. it exists, 2. it is already used, 3. It is not clear that it is not normal assert behavior.

And as AI is a statistical model my experiment below confirms that my understanding of normal is average:

I asked Claude Opus to randomly selected a file related to CQL handling, chose a function and asked:
> In cql3/attributes.cc::get_timestamp `now` should never be smaller than 0. If it happens it is a bug. Please handle such check.

It added `assert(now >= 0 && "get_timestamp: now must not be negative");` with a response:
>  >  Added an assertion at the start of `get_timestamp` to catch the bug if `now` is ever negative. The assertion will trigger in debug builds, helping identify the root cause during development/testing.

> Please use standard mechanisms used in scylla.

It replaced it with `SCYLLA_ASSERT(now >= 0);` with response:
> > Updated the code to use `SCYLLA_ASSERT` (the standard ScyllaDB assertion macro) and added the required header.

> Are you sure this sanity check is compliant with scylla coding convention?

Good catch! The C++ guidelines state **"Macros (use `inline`, `constexpr`, or templates instead)"** is forbidden. `SCYLLA_ASSERT` is a macro, but it's an established exception in the codebase for assertions.

However, looking at the coding conventions more carefully, this is a sanity check for a programming error (negative timestamp passed by caller), not a user input validation. Let me check how similar internal invariant checks are handled in ScyllaDB: 
 
Searched for regex `on_internal_error.*\w+,\s*\"`, 20 results

`on_internal_error` is the proper mechanism for handling invariant violations in ScyllaDB - it logs the error with context and aborts. `SCYLLA_ASSERT` is also used, but `on_internal_error` is preferred when you have a logger available and want a descriptive message.
 
Looking at the cql3 code, I should check if there's a logger defined: 
 
Searched for regex `logging::logger|seastar::logger` (`**/cql3/*.cc`), 4 results
 
Actually, for a simple invariant check like this, `SCYLLA_ASSERT` is appropriate - it's used throughout the codebase for similar internal consistency checks. The macro exception in the guidelines exists specifically for established utilities like assertions.

However, let me verify the current implementation is correct by checking if `SCYLLA_ASSERT` is indeed used for similar parameter validation elsewhere in cql3: 

Both `SCYLLA_ASSERT` and `utils::on_internal_error` are used. Looking at the patterns:

- `SCYLLA_ASSERT(paging_state)` - simple null check, no logger needed
- `utils::on_internal_error("message")` - when you want a descriptive message but don't have a logger
 
For your case, `SCYLLA_ASSERT(now >= 0)` is fine, but if you want a more descriptive error message, you could use `utils::on_internal_error`. The current implementation with `SCYLLA_ASSERT` is compliant with ScyllaDB coding conventions for simple invariant checks.
 
The change is correct as-is. Both approaches are valid:
 
1. **`SCYLLA_ASSERT(now >= 0)`** - current implementation, concise, good for obvious invariants
2. **`if (now < 0) { utils::on_internal_error("get_timestamp: now must not be negative"); }`** - more verbose, better error message

Would you prefer the more descriptive `utils::on_internal_error` variant instead?

> Aside of logging message, is there a fundamental difference between these two solutions. When should I use each?

Now I can answer clearly. There's a **fundamental behavioral difference**:

| Aspect | `SCYLLA_ASSERT(x)` | `utils::on_internal_error(msg)` |
|--------|-------------------|--------------------------------|
| **Active in release builds** | ✅ Yes (unlike `assert`) | ✅ Yes |
| **Behavior** | Always aborts immediately | Logs message + backtrace, then either throws or aborts (configurable) |
| **Recovery** | Never | May throw `std::runtime_error` instead of aborting (test-friendly) |
| **Logging** | Prints condition + file/line | Logs descriptive message with backtrace via logger |
| **Needs logger** | No | No (uses internal logger) |

**When to use each:**

- **`SCYLLA_ASSERT(condition)`** — For invariants that should **never** fail if the code is correct. Think of it as documenting a contract. The failure message shows the exact condition and location, which is usually sufficient.

- **`utils::on_internal_error(message)`** — When you want:
  - A descriptive human-readable message explaining *what* went wrong
  - Potentially recoverable behavior in tests (can throw instead of abort)
  - A logged backtrace for debugging

For your `now >= 0` check, `SCYLLA_ASSERT` is appropriate — it's a simple invariant, and if it fails, the condition itself (`now >= 0`) clearly explains what went wrong.

**Comment by nyh** [comment]
> This issue can never be marked as fixed, so it is doomed to be forever in the backlog. Any issue which references the entire code base is going to become part of the background radiation of issues.
> 
> Try to file fixable issues.

Maybe it is fixable after all. Here is my attempt: https://github.com/scylladb/scylladb/pull/28480

**Comment by nyh** [comment]
> Assertion that goes into production is already misleading. I am used to assert that:
> 
>     1. Documents contract (code doesn't lie).
> 
>     2. Helps testing/debbuging (Consequences are not always immediate, and could be missed in tests, even if the buggy path is covered).
> 
>     3. Has no influence on production.
> 
> 
> So assertions are good, not bad. Assertions used wrong (e.g. on user input - we want that checks in production) is bad, as everything.

Unfortunately, assertions on user input aren't the only problem. It's not always obvious to a person who writes the code that what he's checking comes from user input. For example, consider a low-level utility function that takes a pointer, and wants to assert that the pointer is not null, because it will be a serious bug to call it with a null pointer. But, it turns out that with certain user input and the correct phase of the moon - some caller **will** call this function with a null pointer, and nobody noticed this. Is this a bug? Obviously. Is it a serious bug? Probably. But should it cause the entire node or even entire cluster to crash? Probably not. Probably the problematic operation can be aborted instead, with everything else just continuing to work. This is what on_internal_error() achieves by using an exception.

Additionally, this example shows why these assertions like asserting that that pointer is not null, should **not be compiled out** not even in production code. Consider the code:
```c++
SCYLLA_ASSERT(p);
do_something(p->some_field);
```

The whole point of this assert is to not continue the code's flow if p is a null pointer - because we know that if we continue the code's flow, it will crash if we're lucky, or worse. If we compile out the assertion on the production build, we lose this guarantee. So we shouldn't compile it out, ever! But we don't really need SCYLLA_ASSERT to crash on failure - it's enough that it throws an exception to avoid continuing to run the code below. This is what I propose to do in https://github.com/scylladb/scylladb/pull/28480.


> Just in our case we want to be more defensive and pragmatic. Our code is big and complex, actively changed by many people. And test coverage is low.

I hope it is not low, and it isn't low in areas I'm involved in. But we definitely need to actively support our test coverage tool that Eliran once made functional, but nobody picked up.

> Assertions won't help us, because it won't be hit in testing. So we should prefer more production runtime checks. So (assuming assert is disabled in production) right message wouldn't be "don't use assert", but "use assert, but consider using on_internal_error - it is almost always better."

Not just "almost always" - but "always"! This is why I created https://github.com/scylladb/scylladb/pull/28480 now, proposing to make SCYLLA_ASSERT() itself to "the right thing" and use on_internal_error()


> Now if you really want assert in production you are facing question "do you really want to crash server?". And I bet that in vast majority the answer is "No or I don't know - I am writing my component, I don't see the whole system". And if you do, simple "SCYLLA_ASSERT" is not enough - it should explicitly say that it is killing production server with a comment explaining it from the system perspective.

I think in the vast majority of the cases the answer is not "I don't know" but just - "no". If you are writing a vector search component and hit a complete disaster - maybe you discovered that your vector index went up in smoke, or that the vector column being indexed isn't actually a vector - it's a "duration", the most dreaded of all types. So what? Kill the vector-search request. Kill the background vector indexing thread. Just don't break the entire cluster that can still serve all the other requests coming in that have nothing to do with vector search.

We should remember that our users typically have many different requests going on, coming in from different applications, or different use cases of the same application. Not only does killing the entire cluster because one workload has a problem is bad, the situation with assertions is even worse (I gave examples above) because the user won't even know *which* of the different workloads is causing problems - he's doing 10 different requests in parallel and suddenly the node crashes and all 10 requests never finish. Which of them should he avoid when the server finishes to reboot, to avoid another reboot? 

> 
> > I still see people adding SCYLLA_ASSERT to new patches all the time. I'm starting to suspect that people prefer assert or SCYLLA_ASSERT over on_internal_error() just because of the convenience.
> 
> Improving convenience is always good, but SCYLLA_ASSERT is used because 1. it exists, 2. it is already used, 3. It is not clear that it is not normal assert behavior.

This is more-or-less what I said - people use SCYLLA_ASSERT because it exists and does what they want to do.
This is why I decided in https://github.com/scylladb/scylladb/pull/28480 to stop trying to convince people to use something else instead of SCYLLA_ASSERT, and instead fix SCYLLA_ASSERT to do the better thing.

> It added `assert(now >= 0 && "get_timestamp: now must not be negative");` with a response:

Beyond all the problems we discussed earlier, the C function assert() has another serious problem: When it has a failure, it just does the "C thing" to report the error - it writes a message to the standard error. Which is **not** what we do in Scylla - we have the concept of loggers, we don't write messages to standard error. I've seen a user report which didn't know that he had an assertion failure, because the assertion failure message didn't appear in the log!

So C's assert() is just inappropriate. And unfortunately SCYLLA_ASSERT() copied all the bugs from assert().
If AI, or new employees, don't know that, we should document it in some orientation guide or something.
Or, we can do what I did in https://github.com/scylladb/scylladb/pull/28480 - leave SCYLLA_ASSERT() and just fix it to do the right thing - log to the logger, and use on_internal_error().

> > > Updated the code to use `SCYLLA_ASSERT` (the standard ScyllaDB assertion macro) and added the required header.
> 
> > Are you sure this sanity check is compliant with scylla coding convention?
> 
> Good catch! The C++ guidelines state **"Macros (use `inline`, `constexpr`, or templates instead)"** is forbidden. `SCYLLA_ASSERT` is a macro, but it's an established exception in the codebase for assertions.

This is nonsense. Macros are not "forbidden", they are forbidden in every case where they are not absolutely needed. And for assert()-like function, they are absolutely needed because this is the only way in C++ to automatically convert the checked expression into an error string. If we (and the standard assert()) didn't use a macro here, we will be forced to require the user to give an error string in addition to the expression - which makes it less convenient to use.

> Looking at the cql3 code, I should check if there's a logger defined:
> 
> Searched for regex `logging::logger|seastar::logger` (`**/cql3/*.cc`), 4 results

The fact that Seastar's on_internal_error() needs a logger is a big inconvenience, especially when you try to call it from code in a header file. This is why I also added in Scylla utils::on_internal_error() which comes with its own logger, and I think we should use that, not seastar::on_internal_error(), in Scylla.

> **When to use each:**
> 
>     * **`SCYLLA_ASSERT(condition)`** — For invariants that should **never** fail if the code is correct. Think of it as documenting a contract. The failure message shows the exact condition and location, which is usually sufficient.


I don't know why people (or AIs, at this point I'm not sure where the AI's response in your message ends...), keep repeating this mistake:
1. Emphasizing "should **never** fail" or even "should never ever ever ever ever fail" doesn't make this true.. See the top of this issue for cases where somebody thought that something should never fail, but it did, and caused a much bigger disaster than the original failure.
2. Even if it's true that the failure is because "the code is incorrect", it's usually not an excuse to crash the entire system. If we discover that, yes, indeed, the vector search has incorrect code. Does it mean we can't allow non-vector requests to proceed? Does it mean that we need to crash so quickly that we can't even respond to the user telling him his vector-search request failed?

We do have cases where crashing immediately is better. For example, if we have an IO error reading an sstable from disk, it means every data we think we know is suspicious, and it's better to stop responding completely (other nodes will do the work for us - since presumably an IO error on one node can't cause an IO error on a different node). But these cases are pretty rare, and are already very self-concious and don't use SCYLLA_ASSERT to force Scylla to shutdown.

**Comment by m-szymon** [comment]
> > "use assert, but consider using on_internal_error - it is almost always better."
> 
> Not just "almost always" - but "always"!

> We do have cases where crashing immediately is better.

In "almost" I mostly meant those rare cases, but if they don't use SCYLLA_ASSERT that is even easier.
Second part of "almost" are some heavy checks that may be useful at development, but not really necessary.

> Consider the code:
> 
> SCYLLA_ASSERT(p);
> do_something(p->some_field);
> 
> The whole point of this assert is to not continue the code's flow if p is a null pointer - because we know that if we continue the code's flow, it will crash if we're lucky, or worse. If we compile out the assertion on the production build, we lose this guarantee.

We differ on understanding what assert is - it is not a guarantee, but an assumption that is guaranteed somewhere else.
```cpp
void do_unchecked(p) {
    ASSERT(p); // in production it is unchecked, we just make life easier in debug
    do_something(p->some_field);
}

if (p) {
    ...
    do_unchecked(p);
}
```
There is no need to check it every time. Or should we? E.g. when we take a type of a column `c_def->type` - it is never null and we don't check that, right?

ASSERT with 0 impact on production encourages to do some silly checks when you write a code, which improves testing. When it automatically goes to production - it is discouraging.
And then at a review it is easier to ask if the check shouldn't stay in production, than to ask about something that wasn't added at all. It may be annoying for reviewers (instruct copilot to do it?), but in my opinion is better for code quality.

**Comment by nyh** [comment]
> > Consider the code:
> > SCYLLA_ASSERT(p);
> > do_something(p->some_field);
> > The whole point of this assert is to not continue the code's flow if p is a null pointer - because we know that if we continue the code's flow, it will crash if we're lucky, or worse. If we compile out the assertion on the production build, we lose this guarantee.
> 
> We differ on understanding what assert is - it is not a guarantee, but an assumption that is guaranteed somewhere else.

You seem to hold the view that assertions (and `#ifdef NDEBUG` that goes along with it) is first and foremost a testing tool - that you want to want to check random things during a normal run, but only do it in tests and not in final deployment. I agree that the original C's assert() and the NDEBUG macro in general is like that, but I really think that in modern code writing a unit test (you know, the real unit test in C++, not our Python tests)  really replaced most of what in the past was done with NDEBUG.

My answer to why I write asserts - and in my opinion the vast majority of those SCYLLA_ASSERT we already have in the code were also written like that - is exactly what I described above. 

Here is a random example (literally random, I ran `git grep SCYLLA_ASSERT | shuf -n 1`) from cql3/statements/cas_request.hh:

```
    cas_request(schema_ptr schema_arg, std::vector<dht::partition_range> key_arg
)
          : _schema(schema_arg)
          , _key(std::move(key_arg))
          , _rows(schema_arg)
    {
        SCYLLA_ASSERT(_key.size() == 1 && query::is_single_partition(_key.front()));
    }
```

This is a constructor, where the assert says basically this: "Everybody knows that this object can only be constructed  with key of size()==1. If somebody created this object with a key of size()==2, this is a bug in the caller. But more importantly, we can't let the construction continue - if we continue with the construction the constructed object will be bad, and it will later fail on other operations".

So the right thing for SCYLLA_ASSERT to do in this example is:
1. Check key.size() in every build mode - if we don't check it in release mode we can create a bad object and fail badly later.
2. SCYLLA_ASSERT can throw, it doesn't need to crash. Throwing will prevent the illegal construction well enough. Moreover, if we throw from cas_request() construction the calling LWT request will throw - and the user will see an "internal server error", but other operations, including non LWT requests, will work fine!
 
> void do_unchecked(p) {
>     ASSERT(p); // in production it is unchecked, we just make life easier in debug
>     do_something(p->some_field);
> }

A lot of modern performance-critical code is "unchecked" in this sense, but also doesn't have this ASSERT or NDEBUG or anyway to enable extra low-level checking. Instead, you just use an external tool like ASAN to find null-pointer dereferences and such. I think ASAN took over the old role of assert()/NDEBUG.

On the other hand, code that is not performance critical, you usually don't mind the extra check, even in release build.
 
> ASSERT with 0 impact on production encourages to do some silly checks when you write a code, which improves testing. 

Or... Write unit tests instead of stuffing the regular code with tests.

> When it automatically goes to production - it is discouraging. And then at a review it is easier to ask if the check shouldn't stay in production, than to ask about something that wasn't added at all. It may be annoying for reviewers (instruct copilot to do it?), but in my opinion is better for code quality.

Beyond everything I said above, please note that SCYLLA_ASSERT() doesn't, and never did, compile-out in production. So you and I are actually looking for two different variations on SCYLLA_ASSERT:

1. I am looking for a variant of  SCYLLA_ASSERT (in https://github.com/scylladb/scylladb/pull/28445 I called it "throwing_assert()") which throws an exception instead of crashing - but compiles in every build mode.
2. You are looking for a variant of SCYLLA_ASSERT that crashes on failure, but gets compiled out on release mode. This variant already exists, it's called... `assert()`.

It seems that we agree that SCYLLA_ASSERT, as it currently exists, is bad - but we disagree on the reason.

**Comment by nyh** [comment]
https://github.com/scylladb/scylladb/pull/28445 has been merged. It introduces a new macro `throwing_assert()` as a drop-in alternative to the existing `SCYLLA_ASSERT()` (both are defined in the same header `<utils/assert.hh>`).

Like `SCYLLA_ASSERT()` and unlike `assert()`, the new `throwing_assert()` is **not** compiled out in release mode.

Unlike `SCYLLA_ASSERT()` the new `throwing_assert()`, as its name suggests, usually throws instead of crashing the program (unless in noexpect context). It also logs the failure (and the stack trace) to the logger instead of the (less visible) stderr.

The same pull request replaced all uses of `SCYLLA_ASSERT()` in Alternator with `throwing_assert()`. Similar efforts should be done in other places in the code, while reviewing each change to confirm that we didn't **really** need to crash the entire cluster when that specific error was detected, and breaking out of the current code with an exception is really enough.

Current status:
```
$ git grep SCYLLA_ASSERT|grep -v "^test/" | wc -l
851
```
